File: LSP+ISP/Java/MarsRoverKata/src/main/java/MarsRoverKata/MissionManager.java
package MarsRoverKata;

import MarsRoverKata.Commands.*;
import org.apache.commons.lang3.StringUtils;

import java.util.ArrayList;

public class MissionManager {
    private Rover rover;

    public Rover getRover() {
        return rover;
    }

    public Mars Planet;

    public Mars getPlanet() {
        return Planet;
    }

    private final Commander _commander;

    public MissionManager(Rover rover) {
        this.rover = rover;
        Planet = rover.getMars();
        _commander = new Commander();
        // _commander.CommandExecuted. += updateAliens;
    }

    // TODO: EventHandler implementation
//    private void updateAliens(Object sender, EventArgs e) {
//        Planet.UpdateAliens();
//    }

    public String acceptCommands(String commandString) {
        _commander.setCommands(new ArrayList<ICommand>());

        char[] commands = commandString.toLowerCase().toCharArray();
        boolean success = true;
        for (char command : commands) {
            success = AcceptCommand(success, command);
        }

        return success ? StringUtils.EMPTY : "Some invalid commands were found!!!";
    }

    private boolean AcceptCommand(boolean success, char command) {
        if (command == 'f')
            _commander.accept(new ForwardCommand(rover));
        else if (command == 'b')
            _commander.accept(new BackwardCommand(rover));
        else if (command == 'r')
            _commander.accept(new TurnRightCommand(rover));
        else if (command == 'l')
            _commander.accept(new TurnLeftCommand(rover));
        else if (command == 'm')
            _commander.accept(new FireMissileCommand(rover));
        else if (command == 'g')
            _commander.accept(new FireMortarCommand(rover));
        else
            success = false;
        return success;
    }

    public String ExecuteMission() {
        boolean success = _commander.executeCommands();
        return success ? StringUtils.EMPTY : "An error occured while executing commands";
    }

    public void AddObstacle(int x, int y, String type) {
        Point location = new Point(x, y);
        IObstacle obstacle = CreateObstacle(location, type);
        Planet.addObstacle(obstacle);
    }

    private IObstacle CreateObstacle(Point location, String type) {
        if (type.equalsIgnoreCase("Alien")) {
            return new Alien(Planet, location);
        }
        return new Obstacle(location);
    }
}

File: LSP+ISP/Exercise/MarsRoverKata/MissionManager.cs
﻿using System;
using System.Collections.Generic;
using System.Linq;
using MarsRoverKata.Commands;

namespace MarsRoverKata
{
    public class MissionManager
    {
        public Rover Rover { get; private set; }
        public Mars Planet { get; private set; }
        private readonly Commander _commander;
        
        public MissionManager(Rover rover)
        {
            Rover = rover;
            Planet = rover.Mars;
            _commander = new Commander();
            _commander.CommandExecuted += UpdateAliens;
        }

        private void UpdateAliens(object sender, EventArgs e)
        {
            Planet.UpdateAliens();
        }

        public string AcceptCommands(String commandString)
        {
            _commander.Commands = new List<ICommand>();

            var commands = commandString.ToLower().ToCharArray();
            bool success = true;
            foreach (Char command in commands)
            {
                success = AcceptCommand(success, command);
            }

            return success ? String.Empty : "Some invalid commands were found!!!";
        }

        private bool AcceptCommand(bool success, Char command)
        {
            if (command.Equals('f'))
                _commander.Accept(new ForwardCommand(Rover));
            else if (command.Equals('b'))
                _commander.Accept(new BackwardCommand(Rover));
            else if (command.Equals('r'))
                _commander.Accept(new TurnRightCommand(Rover));
            else if (command.Equals('l'))
                _commander.Accept(new TurnLeftCommand(Rover));
            else if (command.Equals('m'))
                _commander.Accept(new FireMissileCommand(Rover));
            else if (command.Equals('g'))
                _commander.Accept(new FireMortarCommand(Rover));
            else
                success = false;
            return success;
        }

        public string ExecuteMission()
        {
            bool success = _commander.ExecuteCommands();

            return success ? String.Empty : "An error occured while executing commands";
        }

        public void AddObstacle(int x, int y, string type)
        {
            Point location = new Point(x, y);
            var obstacle = CreateObstacle(location, type);
            Planet.AddObstacle(obstacle);
        }

        private IObstacle CreateObstacle(Point location, string type)
        {
            if (type.Equals("Alien", StringComparison.OrdinalIgnoreCase))
            {
                return new Alien(Planet, location);
            }
            return new Obstacle(location);
        }
    }
}
